### DISCUSSION TITLE
Ruby

### DISCUSSION THREAD

**Discussion by ggerganov** [ggml-org/whisper.cpp#507] [Bindings]
General discussion about the `ruby` bindings for `whisper.cpp`

**Current state:** [bindings/ruby](https://github.com/ggerganov/whisper.cpp/tree/master/bindings/ruby)

Initial merge: #500 
CC: @taf2

**Comment by taf2** [comment]
Working on getting a gem published to rubygems so people can just

```
gem install whispercpp
```

**Reply by taf2** [reply]
Looks like someone already published a gem named whisper... So I think if we go with whispercpp - that would make sense and is available.  Will work on pushing a gem and pull request.

**Reply by KitaitiMakoto** [reply]
@taf2 Hi, thank you for publishing the gem. I was excited whisper.cpp bindings already exist.

Are you gem owner of whispercpp gem? Could you check updates(#2484, #2506, #2516, #2518) and publish the new release?

**Reply by taf2** [reply]
hey @KitaitiMakoto i believe i did publish the original gem... happy to publish the updates - i should be able to do this this week.  Let me pull down the changes and confirm if that's cool and should not be a problem

**Reply by KitaitiMakoto** [reply]
Sure! I don't hurry and don't mean to rush you. Thanks for the response.

**Reply by taf2** [reply]
looking into this now we need a new version number to push.

**Reply by KitaitiMakoto** [reply]
How about 1.4.0?

**Reply by justindoody** [reply]
Hopefully this can get published soon. Getting this latest code working was **way** more complicated than I thought it would be (I should have just pushed a forked version to my own rubygems), didn't realize bundler doesn't allow local gems referenced by path (or even git if using a git local config setting) to build native extensions. Referencing directly to the git repo would have worked but this gem is dependent on the rake task moving all sorts of files in order to build which isn't executed when using a direct git reference it seems. Finally got it working in my docker build process by manually building the gem and generating the index `gem generate_index` and hosting a temporary local gem server with `ruby -run -e httpd . -p 9292` in order to get `gem "whispercpp", "1.3.0", source: "http://localhost:9292"` to actually install and build native extensions. Not sure why bundler has all these limitations making it hard to deal with native extensions.

**Reply by taf2** [reply]
@ggerganov I think I need someone to update the gemspec file with a new version so I can push:  

```
diff --git a/bindings/ruby/whispercpp.gemspec b/bindings/ruby/whispercpp.gemspec
index d8f5c0d..641c303 100644
--- a/bindings/ruby/whispercpp.gemspec
+++ b/bindings/ruby/whispercpp.gemspec
@@ -3,8 +3,8 @@ require_relative "extsources"
 Gem::Specification.new do |s|
   s.name    = "whispercpp"
   s.authors = ["Georgi Gerganov", "Todd A. Fisher"]
-  s.version = '1.3.0'
-  s.date    = '2024-05-14'
+  s.version = '1.3.1'
+  s.date    = '2024-12-19'
   s.description = %q{High-performance inference of OpenAI's Whisper automatic speech recognition (ASR) model via Ruby}
   s.email   = 'todd.fisher@gmail.com'
   s.extra_rdoc_files = ['LICENSE', 'README.md']
```

then I can run rake release.... Alliteratively, maybe I can add you to rubygems so I'm not the bottleneck for this?

**Reply by ggerganov** [reply]
I just pushed an update: https://github.com/ggerganov/whisper.cpp/commit/e4e05981d696938894a1308d80d82aef892cd910

Let me know if anything else is needed.

**Reply by KitaitiMakoto** [reply]
@taf2 Ah, you have wanted me to update version number, right? I didn't notice. Sorry.

**Reply by KitaitiMakoto** [reply]
RubyGems.org has [Trusted Publishing](https://guides.rubygems.org/trusted-publishing/) feature. It's worth to consider.

**Reply by taf2** [reply]
@KitaitiMakoto Successfully registered gem: whispercpp (1.3.1)
I'm not familiar with Trusted Publishing but would be great to let someone else push these gems 👍

**Reply by KitaitiMakoto** [reply]
Thanks! I confirmed it is successfully installed and worked on my Mac machine.
ping @justindoody

**Reply by justindoody** [reply]
thanks!

**Reply by KitaitiMakoto** [reply]
@taf2 Hi, can you release v1.3.2 from current `master`?

**Reply by KitaitiMakoto** [reply]
Hi @taf2, I hope you are doing well. Do you have time to release new version? Or, if you are too busy (and trust me), could you consider to give me( https://rubygems.org/profiles/KitaitiMakoto ) a permission to push gem to rubygems.org? Thank you.

**Reply by taf2** [reply]
@KitaitiMakoto should* be live

**Reply by KitaitiMakoto** [reply]
@taf2 Thank you so much!

**Reply by KitaitiMakoto** [reply]
Hi, @taf2 . Could you please release new version? Current `master` includes a fix for installation trouble: https://github.com/ggml-org/whisper.cpp/issues/3259

**Reply by KitaitiMakoto** [reply]
@taf2 Hello, do you have time to cut out new release...?

**Reply by taf2** [reply]
@KitaitiMakoto pushed!

**Reply by KitaitiMakoto** [reply]
@taf2 Thank you!

**Reply by KitaitiMakoto** [reply]
Hi, @taf2 , there is a release request from @adamdebono : https://github.com/ggml-org/whisper.cpp/pull/3365#issuecomment-3378757225
Can you push the new version?

**Reply by taf2** [reply]
pushed gem: whispercpp (1.3.4)

**Reply by KitaitiMakoto** [reply]
Thx!

**Reply by KitaitiMakoto** [reply]
Hi, @taf2 . Another request for release, now. Whisper.cpp v1.8.3 was released. Can you release a new gem version, please?

**Reply by KitaitiMakoto** [reply]
Hi, @taf2 . Could you cut the new version? Whisper.cpp v.1.8.4 was released!

**Reply by taf2** [reply]
will do!

**Reply by KitaitiMakoto** [reply]
I confirmed it's released, thank you!

**Reply by KitaitiMakoto** [reply]
@taf2 Hi, can you push the new version? Current master includes new Whisper.cpp version, Windows support, Parakeet support, etc.

**Reply by taf2** [reply]
@KitaitiMakoto done

**Reply by KitaitiMakoto** [reply]
Thanks, I confirmed installation succeeded!

**Reply by KitaitiMakoto** [reply]
Hi, @taf2 , I fixed some installation issues. I appreciate for you to release the new version when you have time.

**Reply by taf2** [reply]
@KitaitiMakoto "Successfully registered gem: whispercpp (1.3.8)"

**Reply by adamdebono** [reply]
Is there a changelog or version history specific to the Ruby gem? I couldn't see anything in the repo, and the version numbers don't align with the [GitHub releases](https://github.com/ggml-org/whisper.cpp/releases).

It would be useful to know what has been changed with respect to the Ruby binding, and how it lines up with the main release versions.

**Reply by taf2** [reply]
Good idea - I’ll see if I can put one together for the next release

**Reply by KitaitiMakoto** [reply]
Thank you, @taf2 !

**Reply by taf2** [reply]
ok just need this PR approved and merged: https://github.com/ggml-org/whisper.cpp/pull/3959

**Reply by KitaitiMakoto** [reply]
Great! I will take a look later.

**Reply by KitaitiMakoto** [reply]
I've done. Thanks.

**Comment by taf2** [comment]
@ggerganov I'm looking for good solution to interrupt whisper_full_parallel. In ruby we can wrap the call to this in an unlock method, since ruby has a global interpreter lock it is ideal to unlock while the c++ code runs allowing ruby to run other threads and even be interrupted via ctrl+c and other interrupt signals.   It looks whisper_full_parallel calls whisper_full in each thread or just calls whisper_full, if there was a maybe volatile stop variable we added to the context then we could toggle that to false/0 and cause the while loop to break.  thoughts?

**Reply by ggerganov** [reply]
There is currently an existing abort mechanism via the "encoder_begin" callback:

https://github.com/ggerganov/whisper.cpp/blob/f420de1322a64b9ed8cc8705ad0a3fdb9998303a/whisper.h#L240-L245

Would this work?

**Reply by taf2** [reply]
ah perfect - i'll dig into this and rework the ruby bindings to use that shortly.

**Reply by ggerganov** [reply]
Usage of the abort mechanism is demonstrated here:

https://github.com/ggerganov/whisper.cpp/blob/b4ebdb6b57dedf9e82eec6e13ac70e8089e06190/examples/main/main.cpp#L603-L616

**Comment by brauliobo** [comment]
am I missing something? It doesn't install here. I have `whisper.cpp` AUR package installed on the system

```
braulio @ bhavapower ➜  ruby-whisper.cpp git:(main) ✗  LANG=en gem install whispercpp
Building native extensions. This could take a while...
ERROR:  Error installing whispercpp:
        ERROR: Failed to build gem native extension.

    current directory: /home/braulio/.rvm/gems/ruby-3.1.2/gems/whispercpp-1.3.0/ext
/home/braulio/.rvm/rubies/ruby-3.1.2/bin/ruby -I /home/braulio/.rvm/rubies/ruby-3.1.2/lib/ruby/3.1.0 -r ./siteconf20240926-2567174-n68uhi.rb extconf.rb
cp: cannot stat './../../../whisper.cpp': No such file or directory
cp: cannot stat './../../../whisper.h': No such file or directory
cp: cannot stat './../../../ggml.h': No such file or directory
cp: cannot stat './../../../ggml.c': No such file or directory
cp: cannot stat './../../../ggml-impl.h': No such file or directory
cp: cannot stat './../../../ggml-alloc.h': No such file or directory
cp: cannot stat './../../../ggml-alloc.c': No such file or directory
cp: cannot stat './../../../ggml-backend-impl.h': No such file or directory
cp: cannot stat './../../../ggml-backend.h': No such file or directory
cp: cannot stat './../../../ggml-backend.c': No such file or directory
cp: cannot stat './../../../ggml-common.h': No such file or directory
cp: cannot stat './../../../ggml-quants.h': No such file or directory
cp: cannot stat './../../../ggml-quants.c': No such file or directory
cp: cannot stat './../../../examples/dr_wav.h': No such file or directory
creating Makefile

current directory: /home/braulio/.rvm/gems/ruby-3.1.2/gems/whispercpp-1.3.0/ext
make DESTDIR\= clean

current directory: /home/braulio/.rvm/gems/ruby-3.1.2/gems/whispercpp-1.3.0/ext
make DESTDIR\=
compiling ggml.c
ggml.c:4:10: fatal error: ggml-impl.h: No such file or directory
    4 | #include "ggml-impl.h"
      |          ^~~~~~~~~~~~~
compilation terminated.
make: *** [Makefile:246: ggml.o] Error 1

make failed, exit code 2

Gem files will remain installed in /home/braulio/.rvm/gems/ruby-3.1.2/gems/whispercpp-1.3.0 for inspection.
Results logged to /home/braulio/.rvm/gems/ruby-3.1.2/extensions/x86_64-linux/3.1.0/whispercpp-1.3.0/gem_make.out
```

**Reply by oriolgual** [reply]
I am also getting this error, how did you fix it?

**Reply by KitaitiMakoto** [reply]
This should be fixed by #2484

**Reply by justindoody** [reply]
@KitaitiMakoto why didn't you bump the gem version so there would be a new release on rubygems?

**Reply by KitaitiMakoto** [reply]
Because I thougt the gem owner will do it just before releasing the new version.

**Reply by KitaitiMakoto** [reply]
Now it's released.

**Comment by brauliobo** [comment]
just pushed a new gem at https://rubygems.org/gems/whisper.cpp

**Reply by oriolgual** [reply]
Hi! Can this be installed without CUDA? I'm getting this error:

```
Cloning whisper.cpp repository...
Cloning into 'whisper.cpp'...
Building libwhispercpp.so with GGML_CUDA=1...
I ccache not found. Consider installing it for faster compilation.
expr: syntax error
I whisper.cpp build info:
I UNAME_S:   Darwin
I UNAME_P:   arm
I UNAME_M:   arm64
I CFLAGS:    -Iggml/include -Iggml/src -Iinclude -Isrc -Iexamples -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DGGML_USE_BLAS -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_CUDA -I/usr/local/cuda/include -I/usr/local/cuda/targets/arm64-linux/include  -std=c11   -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Werror=implicit-function-declaration -pthread -Wunreachable-code-break -Wunreachable-code-return -Wdouble-promotion
I CXXFLAGS:  -std=c++11 -fPIC -O3 -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wmissing-declarations -Wmissing-noreturn -pthread   -Wunreachable-code-break -Wunreachable-code-return -Wmissing-prototypes -Wextra-semi -Iggml/include -Iggml/src -Iinclude -Isrc -Iexamples -D_XOPEN_SOURCE=600 -D_DARWIN_C_SOURCE -DNDEBUG -DGGML_USE_ACCELERATE -DGGML_USE_BLAS -DACCELERATE_NEW_LAPACK -DACCELERATE_LAPACK_ILP64 -DGGML_USE_CUDA -I/usr/local/cuda/include -I/usr/local/cuda/targets/arm64-linux/include
I NVCCFLAGS: -std=c++11 -O3 -use_fast_math --forward-unknown-to-host-compiler -arch=native -DGGML_CUDA_DMMV_X=32 -DGGML_CUDA_MMV_Y=1 -DK_QUANTS_PER_ITERATION=2 -DGGML_CUDA_PEER_MAX_BATCH_SIZE=128
I LDFLAGS:   -framework Accelerate -lcuda -lcublas -lculibos -lcudart -lcublasLt -lpthread -ldl -lrt -L/usr/local/cuda/lib64 -L/usr/lib64 -L/usr/local/cuda/targets/arm64-linux/lib -L/usr/local/cuda/lib64/stubs -L/usr/lib/wsl/lib
I CC:        Apple clang version 16.0.0 (clang-1600.0.26.3)
I CXX:       Apple clang version 16.0.0 (clang-1600.0.26.3)
/bin/sh: nvcc: command not found
I NVCC:
/bin/sh: nvcc: command not found
grep: invalid option -- P
usage: grep [-abcdDEFGHhIiJLlMmnOopqRSsUVvwXxZz] [-A num] [-B num] [-C[num]]
	[-e pattern] [-f file] [--binary-files=value] [--color=when]
	[--context[=num]] [--directories=action] [--label] [--line-buffered]
	[--null] [pattern] [file ...]
Makefile:855: *** I ERROR: For CUDA versions < 11.7 a target CUDA architecture must be explicitly provided via environment variable CUDA_DOCKER_ARCH, e.g. by running "export CUDA_DOCKER_ARCH=compute_XX" on Unix-like systems, where XX is the minimum compute capability that the code needs to run on. A list with compute capabilities can be found here: https://developer.nvidia.com/cuda-gpus .  Stop.
Failed to build libwhisper.a and libggml.a

extconf failed, exit code 1
```

**Comment by brauliobo** [comment]
@oriolgual unfortunately not, currently `GGML_CUDA=1` is set on the `extconf.rb`. I'm not sure it is a good idea to allow the user to set this environment variable on the build.

**Reply by oriolgual** [reply]
And why is CUDA required?

**Reply by brauliobo** [reply]
it is just my use case here. would be ideal if whisper.cpp could be compiled with everything and then cuda is selected on runtime

**Comment by KitaitiMakoto** [comment]


**Reply by ggerganov** [reply]
`ggml-backend.c` is old and should be removed. It was changed to C++ `ggml-backend.cpp`

**Reply by KitaitiMakoto** [reply]
@ggerganov 
Oops. I removed my comment because it should be on Issue. I filed my removed comment as an issue: https://github.com/ggerganov/whisper.cpp/issues/2483
I'm sorry for removing it.

BTW, thanks for your quick response.

**Comment by KitaitiMakoto** [comment]
I would like to make GPU work, especially on my Mac machine (Apple silicon, Metal). But, I need learn many things such as building C/C++, Make, assembly, GPU (CUDA, Metal, ...), GGML and so on. For instance, I cannot understand Makefile to build examples/main/main.cpp. Learning takes time. Could someone work with GPU? Or, show me brief direction of learning?

I found a document telling how to set some build flags: https://docs.ruby-lang.org/en/3.3/extension_rdoc.html#label-Prepare+extconf.rb

**Reply by brauliobo** [reply]
The gem I did install with cuda enabled, just run gem install whisper.cpp

**Reply by KitaitiMakoto** [reply]
I will check your code and learn how-to. Thanks!

**Reply by KitaitiMakoto** [reply]
Achieved by #2516. Your gem let me know how-to. Thanks!

**Comment by KitaitiMakoto** [comment]
Now whispercpp gem cannot be installed on Windows. I believe it's possible, but I cannot develop for Windows because I don't have Windows environment. Anybody can do that?

**Reply by taf2** [reply]
pushed gem: whispercpp (1.3.4)

**Reply by KitaitiMakoto** [reply]
v1.3.7 installs on Windows!

**Comment by taf2** [comment]
pushed 1.3.5

**Reply by KitaitiMakoto** [reply]
Thank you!
